src/main/java/com/elasticpath/tools/smcupgrader/GitClient.java

```diff
@@ -195,26 +204,70 @@ public Set<IndexEntry> getStatusIndexEntries() {
 	}
 
 	/**
-	 * Checks the commit history for the file represented by the given change, and determines whether every commit can also be found on the
-	 * given remote repository.
+	 * Returns an iterable of commits containing the specified path in any branch with a name starting with upstreamRemoteName.
 	 *
-	 * @param change             the change
-	 * @param upstreamRemoteName the name of the upstream remote to check for commits
-	 * @return {@code true} if every commit for the file represented by the given change exists on the remote repository
+	 * @param path a path that the commit must contain
+	 * @param upstreamRemoteName the branch name prefix
+	 * @return an iterable of commits
 	 */
-	public boolean allLocalCommitsExistInRemote(final Change change, final String upstreamRemoteName) {
+	public Iterable<RevCommit> getAllCommitsForPathInAllBranches(final String path, final String upstreamRemoteName) {
 		try (Git git = new Git(repository)) {
-			final Iterable<RevCommit> commits = git.log()
-					.addPath(change.getPath())
-					.call();
+			// Get all branches that start with "refs/remotes/" + upstreamRemoteName + "/"
+			String remoteBranchPrefix = "refs/remotes/" + upstreamRemoteName + "/";
+			List<AnyObjectId> upstreamRemoteBranchHeads = git.branchList()
+					.setListMode(ListBranchCommand.ListMode.REMOTE)
+					.call()
+					.stream()
+					.filter(ref -> ref.getName().startsWith(remoteBranchPrefix))
+					.map(ref -> (AnyObjectId) ref.getObjectId())
+					.collect(Collectors.toList());
+
+			if (upstreamRemoteBranchHeads.isEmpty()) {
+				return Collections.emptyList(); // No matching branches
+			}
 
-			return StreamSupport.stream(commits.spliterator(), true)
-					.allMatch(revCommit -> commitExistsInRemote(git, upstreamRemoteName, revCommit));
+			LogCommand logCommand = git.log()
+					.addPath(path);
+			for (AnyObjectId branchHead : upstreamRemoteBranchHeads) {
+				logCommand.add(branchHead);
+			}
+			return logCommand.call();
+		} catch (final GitAPIException | IncorrectObjectTypeException | MissingObjectException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	/**
+	 * Returns an iterable of commits in the current branch that contain the specified path.
+	 *
+	 * @param path a path that the commit must contain
+	 * @return an iterable of commits
+	 */
+	public Iterable<RevCommit> getAllCommitsForPath(final String path) {
+		try (Git git = new Git(repository)) {
+			return git.log()
+					.addPath(path)
+					.call();
 		} catch (final GitAPIException e) {
 			throw new RuntimeException(e);
 		}
 	}
 
+	/**
+	 * Returns the latest commit in the current branch that contains the specified path.
+	 *
+	 * @param path a path that the commit must contain
+	 * @return a commit
+	 */
+	public RevCommit getLatestCommitForPath(final String path) {
+		Iterator<RevCommit> iterator = getAllCommitsForPath(path).iterator();
+		if (iterator.hasNext()) {
```

Comment:
Seems costly to call `getAllCommitsForPath` when you only ever want the first.  Perhaps there a saving here?

Reply:
Possibly, yeah, let me look into that.

Reply:
Fixed.